### ISSUE TITLE
ResponsiveContext loading results in FOUC

### ISSUE THREAD

**Issue by sergeylukin** [grommet/grommet#3496]
### Expected Behavior

Grommet provide API for listening to responsive context bootstrapping

### Actual Behavior

Grommet responsive process is using `medium` size by default, and it takes time for it to respond to actual screen size and re-render components which results in `flashing` effect when grid may have 2 columns for a second or two and then it's replaced with 1 column grid on small screen.

### URL, screen shot, or Codepen exhibiting the issue

Open this on mobile: https://yay.tips

### Steps to Reproduce

1. Have a grid like following:

```
    <ResponsiveContext.Consumer>
        {size => {
          const columns = {
            small: '',
            medium: ['1/2', '1/2'],
            large: ['1/2', '1/2'],
          }
          return (
            <Box pad={{ vertical: 'large' }}>
                <Grid
                  align="start"
                  columns={columns[size]}
                  gap="large"
                >
                  ...
                </Grid>
            </Box>
          )}}
      </ResponsiveContext.Consumer>
```

2. Open on mobile screen, instead of seeing 1 column as soon as page is loaded, you will see 2 columns grid for a second and only then 1

### Your Environment

- Grommet version: 2
- Browser Name and version: any
- Operating System and version (desktop or mobile): any

**Comment by TheWeeezel** [comment]
This is an issue which is not related to grommet. It's a technical issue. You can read about it in the gatsby issues section. It's not even an issue considered to the gatsby team. I spent quite some time trying to fix this. 
From what I understood is that gatsby renders the page without knowing the viewport of the device it displays towards. so only after receiving back the information the correct viewport is set ergo the breakpoints. Since this will always be the case it must interpret a default value. You can override the default setting. Which means either show always the small viewport first or the large viewport. Or you can introduce a splash screen. Which is what I did. See below.

Please check here:
https://stackoverflow.com/questions/57188179/show-overlay-only-once-on-page-entrance-not-route-change-howto

I have now introduced a small splash screen long enough to load the correct viewport.

**Comment by sergeylukin** [comment]
@TheWeeezel thank you very much for response! That information seems very helpful and relevant. I will definitely try it!
It's a pity Grommet's team isn't responding.

**Comment by oorestisime** [comment]
Hey hey sorry for the long wait. indeed this is a limitation for SSR which is unrelated to grommet. This particular problem is mostly related to gatsby since during build time it doesn't have any way of knowing the view port. 
Not sure if there's anything that can be done in grommet for that matter. Let us know or feel free to close this :)

**Comment by halocline** [comment]
FYI, the FOUC is related to styled-components and can be eliminated from Gatsby by using this plugin: https://www.gatsbyjs.org/packages/gatsby-plugin-styled-components/

Basically you need to add `gatsby-plugin-styled-components` and `babel-plugin-styled-components` as dependencies, then modify `gatsby-config.js` to apply the plugin.

These posts provide additional background context as well:
- https://stackoverflow.com/questions/50155946/why-do-we-need-to-install-gatsby-plugin-styled-components-after-installed-styled/51571008#51571008
- https://github.com/gatsbyjs/gatsby/issues/5667#issuecomment-410792495

**Comment by ShimiSun** [comment]
Thanks for sharing @halocline ! 
@sergeylukin I'll be closing this issue, feel free to reopen or ask more questions.

**Comment by halocline** [comment]
For clarification, the plugin above resolves the unstyled content flash.  The responsive context portion @TheWeeezel and @oorestisime mention is likely a separate issue where Gatsby is building the project with an assumed viewport --> initial render occurs --> responsive context --> re-render occurs, which could result in some jumpiness. We'd need to investigate further, but I'd be curious if one could delay Grommet from rendering until window is available? Something like `typeof(window) !== 'undefined' && <Grommet> ....`

**Comment by ShimiSun** [comment]
Thanks, I've reopened to leave room for discussion, @sergeylukin feel free to give this suggestion a try and let us know if that works.

**Comment by oorestisime** [comment]
This is not a good idea since performance and SEO would be heavily impacted. But even if someone would do this it would happen outside of grommet context. 
In some of the application i solved this by delaying some parts of the application not all of it and trying to heavily use media queries rather than JS for some of the component structure.

**Comment by sergeylukin** [comment]
Thanks everyone for jumping in and sharing your opinions. @ShimiSun thanks for keeping it open until we reach an end-of brainstorming or come up with a satisfying solution here. This weekend I will try different methods and will get back with the results of my research. Stay tuned 😉

**Comment by Bahlaouane-Hamza** [comment]
It's a trade-off you need to work with. What you need to do is ship your components with media queries from the beginning.

Let say you have 
`<Box  className="box" basis="1/2" />`

What you get is 
`.box { flex-basis: 50% }`

There is an interesting problem here. How might we represent a component that gets styled differently at different breakpoints? Let’s imagine a responsive context example.)

`export default function App() {
   const size = useContext(ResponsiveContext)
  return (
    <Grommet theme={customTheme}>
      <Box direction="row" fill="horizontal" wrap={true}>
        <Box background="dark-1" basis={size === "small" ? "full" : "1/2"} pad="medium"></Box>
        <Box background="accent-1" basis={size === "small" ? "full" : "1/2"} pad="medium"></Box>
      </Box>
    </Grommet>
  );
}`

Using an SSG like Gatsby will build your HTML based on a certain viewport as @halocline explained.
At the end you get that jumping effect.

As a front-end web developper you always need to hide or show stuff in mobile when you do responsive design.
If you used Bootstrap before. you could easily use their Grid system or utilities to hide stuff. It's still a possibility to mix Grommet with Bootstrap.

You could also look at https://github.com/artsy/fresnel approach.  

We are currently working with new approaches, and we are still figuring it out. my idea here is to go back a little bit in time. back when your HTML markup would contain a component for desktop and another for mobile. But instead of duplicating the markup, right now we have much control. So we only duplicate CSS rules. **SHIPPING COMPONENTS WITH THEIR MEDIA QUERIES !**

I came out with a POC:  https://codesandbox.io/s/serene-kowalevski-tu4np

![grommet-responsive-poc](https://user-images.githubusercontent.com/5018118/72642342-66131800-396c-11ea-806b-071a33132c1f.png)

Your box component has all the media queries it need now. on mobile flex-basis:100%; on medium it's 50%, and the large it's equal to: 75%.

But that's too much complicated to ask for. we will need a complete re-write of grommet in order for all props to accept things like this. Say i wanna change gap in mobile ? pad in tablet.... the list is long enough.

You get the idea. Use styled component to extend and put CSS rules directly there. I'm still new to Grommet internals so i'm not sure if this not going to break things up.
 
What i'm looking for in term of responsive design using Grommet is mainly 3 things.

1. I can access breakpoints from styled components
2. I can quickly hide or show stuff. Say i wanna hide `<MyComponent />` in mobile & tablet `<MediaQuery hide={["small","medium"]}><MyComponent /></MediaQuery>`
3. Something like to POC above (which will prevent me from writing custom styled components) `<Box
          background="dark-1"
          basis={
    sizes: [
      { breakpoint: "large", basis: "3/4" },
      { breakpoint: "medium", basis: "1/2" },
      { breakpoint: "small", basis: "full" }
    ]
  }
          pad="medium"
        />`

I'm very new to React, Grommet & styled components so forgive my naming convention and understanding of how things work.  I'm just trying to explain an idea here because it's a must for me, so bear with me on that. I love how Grommet pushes you to composition over inheritance !

**Comment by Bahlaouane-Hamza** [comment]
@ericsoderberghp any thoughts or guidance on this ? 
Rebass JS is providing a similar approach as the above.
https://rebassjs.org/props#responsive-styles

Thank you for your time.

**Comment by oorestisime** [comment]
Hi,

thanks for taking the time to write this and being patient with us!
You provided really nice examples and i agree that grommet should provide a way of embedding those media queries in the component.

To answer your points:

1. this is already possible, they are in the theme. unless i didn't understand what you meant.
2. This could be a nice utility prop. It would avoid a lot of userland code doing
```
const size = useContext(ResponsiveContext)
if (size === "small") {
return A
}else if (size === "medium") {
 return B
} else {
return C
}
```

I am not entirely sure there is point in adding such prop in every component, I can see why it is helpful in a Box but heading paragraph button etc i think not. In those components your solution is  probably to alter some font size margin padding rather than completly hiding it

3. Have been working with styled-system a few weeks now and i must say the array approach for responsive work very very good. i think it could be of great value for users if they could provide arrays for some props that behave based on the responsive breakpoints

i.e
instead of doing

```
const size = useContext(ResponsiveContext)
return <Text size={size === "small" ? "small" : "large"} /> 
```

it would be something like `<Text size={["small", "large"]}`. Those could also be with media queries. 

I could see great value for this for margin, padding, size already and probably into some other props. 
What's nice with this is we could do it as in incremental approach keeping backwards compatibility. If user provides single value or object then we do this as now otherwise we provide breakpoints for it.


The only downside with this is that grommet kind of does a few things for responsive text already such as reducing the fontsize based on a breakpoint etc. Not sure how this would work with it but worth exploring.

I am interesting in taking a go at this and exploring some solutions. @ShimiSun @ericsoderberghp thoughts?

**Comment by justdanallen** [comment]
I have used Grommet fairly heavy on few Gatsby projects now. I have almost decided switch from Gatsby to Theme-UI/Rebass because of how those libraries handle responsive vs Grommet. The main reason is not the API, but the use of media queries for responsive styles in those libraries. Grommets responsive context is useless for Gatsby SSR builds, because of the before mentioned unknown viewport at SSR. ([Grommet defaults to tablet](https://github.com/grommet/grommet/blob/f4611c750e297e560d35a6403f2eae41a20a4ad7/src/js/components/Grommet/Grommet.js#L127)) .

The use cases I've ran into with this is where, the layout is different for mobile vs desktop. Like hiding things on mobile, reorienting the layout on desktop/mobile, or a mobile vs desktop nav bar. For example a mobile nav vs desktop. Currently with Grommet and ResponsiveContext, on mobile the desktop nav will flash for a second until Grommet mounts and updates the ResponsiveContext based off the current viewport. (it is previously set to medium during ssr). 

@oorestisime Your utility prop example uses ResponsiveContext which wouldn't help the problem with the SSR on gatsby. It needs to be based off of media queries. 

Some current work arounds:

Emotion's [facepaint](https://github.com/emotion-js/facepaint**) -> You could add a facepaint instance on the theme that was configured to the breakpoints.
[React Media](https://github.com/ReactTraining/react-media)

I like the facepaint API a bit better, IMO. 

I have also built my own HOC that would wrap a Grommet component and add a mediaQuery prop, to add styles. I did this before I found out about facepaint....

[Code Sanbox - withMediaQuery demo](https://codesandbox.io/s/grommet-withmediaquery-example-iyyeo)



Some other Wishes:
* I really like the styled-system theme API for responsive (Theme-ui/Rebass/Chakra are based of this). You can do an array [] or {}. It's well thought out and at least worth [reading](https://styled-system.com/responsive-styles). Also note, the object lets you define a default `<Box width={{ _: 1, sm: 1, md: 1 / 2, lg: 1 / 4 }} />`.
* I also like the shorthands for css properties with a top, right, bottom, left. Ex: margin-left = ml <Box ml="small" />.

**Comment by ShimiSun** [comment]
Thanks for sharing your inputs @justdanallen.
@oorestisime I'd love to hear your inputs here.

**Comment by michaelplazek** [comment]
As noted above, this really just comes down to the mechanisms of static site generation and how/when the JS is executed. For what its worth, I've solved this problem in the past in the following way, although I don't believe it's ideal given the potential impact on SEO:

1. Add an id to the `body` in the `html.js` file in the root of your `src`, for example:
```
export default function HTML(props) {
  return (
    <html {...props.htmlAttributes}>
      <head>
        <meta charSet="utf-8" />
        <meta
          name="viewport"
          content="width=device-width, initial-scale=1, shrink-to-fit=no"
        />
        <meta httpEquiv="x-ua-compatible" content="ie=edge" />
        {props.headComponents}
      </head>
      <body id="hide-content" {...props.bodyAttributes}>
        {props.preBodyComponents}
        <div
          key={`body`}
          id="___gatsby"
          dangerouslySetInnerHTML={{ __html: props.body }}
        />
        {props.postBodyComponents}
      </body>
    </html>
  )
}
```
2. Use the lifecycle hook `onInitialClientRender` in your `gatsby-browser.js` to wait until the JS loads before showing the content:
```
/**
 * Since the HTML is prerendered, need to wait until the JS
 * has loaded (and the styles are applied) before
 *  showing the content.
 */
export const onInitialClientRender = () => {
  const item = document.getElementById("hide-content")
  item.removeAttribute("id")
}
```

The white screen is so brief and unnoticeable that I prefer that to the flash of a loader or splash screen. I believe the only way to fully this solve issue for SSR & SSG apps would be to rely on a non-JS mechanism for the styling.

**Comment by jcfilben** [comment]
Going to go ahead and close this, feel free to reopen or comment if continued discussion is needed on this.
